src/Logging/HTTPOutputHandler.php

```diff
@@ -145,6 +146,14 @@ protected function write(array $record)
     {
         ini_set('display_errors', 0);
 
+        // Don't output deprecation warnings to the browser/CLI - rely on logs for those.
```

Comment:
Browser I can understand due to the sheer ugliness of it.  Not so sure I agree with CLI though, I think I'd want to see at very least E_DEPRECATED, though also probably also E_USER_DEPRECATED

Reply:
I'll affect all `E_DEPRECATED` and `E_USER_DEPRECATED` errors. I don't know of any mechanism to check if a given `E_USER_DEPRECATED` error is emitted by us explicitly.

Reply:
To clarify, what do you want to happen in the following specific scenarios?
1. `E_DEPRECATED` error thrown during CLI execution
1. `E_DEPRECATED` error thrown during browser execution
1. `E_USER_DEPRECATED` error thrown during CLI execution
1. `E_USER_DEPRECATED` error thrown during browser execution

And so we want some methods on `Deprecation` to allow configuration of any of those behaviours?

Reply:
E_DEPRECATED is for PHP language depredations, I don't think we should be touching those this what we're trying to do here is improve the user experience of upgrading between Silverstripe majors, not php versions

We should be able to differentiate between the E_USER_DEPRECATED warnings thrown by us vs 3rd party modules like symfony, because we use Deprecation::notice() to record our own. 

I'm not quite sure why were even using HTTPOutputHandler here?  If we want to not show output our own deprecations warnings, then shouldn't we just not output them from the Deprecation class?

In terms of outputting to browser/cli - seems like we could have some config available to toggle them on/off - I'd probably have browser off by default and CLI on by default.  I think seeing a wall of deprecation notices in CLI when you run unit tests is fine since you would have consciously togged on Deprecation::enable()

Reply:
> E_DEPRECATED is for PHP language depredations, I don't think we should be touching those this what we're trying to do here is improve the user experience of upgrading between Silverstripe majors, not php versions

I am going to infer from this that you mean "For the first two scenarios involving `E_DEPRECATED`, we should just let those be handled the way they are already being handled. This PR should not affect how those are being output."

> We should be able to differentiate between the E_USER_DEPRECATED warnings thrown by us vs 3rd party modules like symfony, because we use Deprecation::notice() to record our own.

.... Right, but we throw them using `user_error` which is just an alias of [`trigger_error`](https://www.php.net/manual/en/function.trigger-error.php) which doesn't let us add any context to indicate that this is _our_ deprecation message vs anyone else's. By the time we're handling them in logging or outputting in `HTTPOutputHandler` we don't have the context that _we_ sent it, it's just a `E_USER_DEPRECATED` same as the rest.

> I'm not quite sure why were even using HTTPOutputHandler here? If we want to not show output our own deprecations warnings, then shouldn't we just not output them from the Deprecation class?

I'm not about to fully refactor how we're outputting errors. A `E_USER_DEPRECATED` error triggered via `user_error` will appear in the logs (which we want) _and_ it will be picked up by `HTTPOutputHandler` to determine _if and how_ it should be output in the response. That's the way it works. That's the way it has always worked, and I'm not going to change that.

> In terms of outputting to browser/cli - seems like we could have some config available to toggle them on/off - I'd probably have browser off by default and CLI on by default. I think seeing a wall of deprecation notices in CLI when you run unit tests is fine since you would have consciously togged on Deprecation::enable()

Okay. I'll do that. That's all I wanted to know.

Reply:
> I am going to infer from this that you mean "For the first two scenarios involving E_DEPRECATED, we should just let those be handled the way they are already being handled. This PR should not affect how those are being output."

This PR has this code
```php
            if ($errorCode === E_USER_DEPRECATED || $errorCode === E_DEPRECATED) {
                return;
            }
```
Seems like it shouldn't have `|| $errorCode === E_DEPRECATED` ?  Otherwise we're stopping those from reaching the front end

> I'm not quite sure why were even using HTTPOutputHandler here? If we want to not show output our own deprecations warnings, then shouldn't we just not output them from the Deprecation class?

Does user_error() go "straight" to HTTPOutputHandler::write()? or does a bunch of user_error() get "buffered" and all arrive at HTTPOutputHandler::write() all together?

Just wondering if we could do something like have a public static on Deprecation e.g. Deprecation::IAmOutputtingSilverstripeEUserDeprecatedNow that's set before/after the while loop in Deprecation::outputNotices() that's  readable from HTTPOutputHandler::write().  Use that to toggle the 'continue' statement.  That way we could isolate the E_USER_DEPRECATED to Silverstripe ones and let the Symfony ones pass through

Reply:
> Seems like it shouldn't have || $errorCode === E_DEPRECATED ? Otherwise we're stopping those from reaching the front end

This sounds like a confirmation that my inference is correct and you'd prefer for the first two scenarios (i.e. dealing with `E_DEPRECATED`) to just continue working as it did before this PR. 👍 I'm happy to make that change.
If instead you're asking for some clarification about something.... please ask a direct question so I know what you're asking.

> Does user_error() go "straight" to HTTPOutputHandler::write()? or does a bunch of user_error() get "buffered" and all arrive at HTTPOutputHandler::write() all together?

No, and no. `user_error()` just immediately throw out a `E_USER_*` error. In the case of our deprecation warnings it's `E_USER_DEPRECATED`. The `HTTPOutputHandler` is a monolog handler. Monolog monitors for errors, and it passes them down to its handlers. It has no context about the fact that we threw it vs some other random vendor throwing it. But we're getting _way_ too far into the weeds here. I don't think this is a discussion material to the actual behavioural changes we're making here.

> Just wondering if we could do something like have a public static on Deprecation e.g. `Deprecation::IAmOutputtingSilverstripeEUserDeprecatedNow` that's set before/after the while loop in `Deprecation::outputNotices()` that's readable from `HTTPOutputHandler::write()`

..... I guess we could do that. And I'll do it if you push for me to do it. But that feels like more complexity than the acceptance criteria on the card are asking for.

Reply:
> But that feels like more complexity than the acceptance criteria on the card are asking for.

ACs aren't don't always catch everything. The intention of the card is assist developers upgrading Silverstripe major versions, and we exclusively use the Deprecation class for emitting the Silverstripe deprecations.  Symfony deprecations will show regardless whether or not you have run Deprecation::enable() in config.  We shouldn't be altering the behaviour of Symfony depredations if we can avoid it

Reply:
Okay. If it's all done synchronously (i.e. if I can use a static wrapper around `user_error()` as you've suggested) I'll do that. If it's not immediately doable that way I'm not going to spend more time on it though.

Reply:
You never directly responded to my list of four scenarios... but here's what I've inferred your preference is based on the comments above:

1. `E_DEPRECATED` error thrown during CLI execution
    - No change made in this PR - just let it do whatever it was doing before.
1. `E_DEPRECATED` error thrown during browser execution
    - No change made in this PR - just let it do whatever it was doing before.
1. `E_USER_DEPRECATED` error thrown during CLI execution
    - By default, these should be output. There should be some configuration option to disable that output.
    - If the deprecation warning was _not_ thrown from the `Deprecation` class in framework, just let it do whatever it was doing before this PR.
1. `E_USER_DEPRECATED` error thrown during browser execution
    - By default, these should _not_ be output. There should be some configuration option to enable that output.
    - If the deprecation warning was _not_ thrown from the `Deprecation` class in framework, just let it do whatever it was doing before this PR.

That's what I've done now. In keeping with the fact that enabling warnings has both a static method and an env variable, I've added both for these new options as well.

Reply:
Yup, those 4 points look good, thanks